neuralmonkey/decoders/beam_search_decoder.py

```diff
@@ -248,65 +270,84 @@ def body(*args) -> BeamSearchLoopState:
             logprobs = unfinished_logprobs + finished_logprobs
 
             # update hypothesis scores
-            # shape(hyp_probs) = beam x vocabulary
+            # shape(hyp_probs) = (batch * beam) x vocabulary
             hyp_probs = tf.expand_dims(bs_state.logprob_sum, 1) + logprobs
 
             # update hypothesis lengths
             hyp_lengths = bs_state.lengths + 1 - tf.to_int32(bs_state.finished)
 
-            # shape(scores) = beam x vocabulary
+            # shape(scores) = (batch * beam) x vocabulary
             scores = hyp_probs / tf.expand_dims(
                 self._length_penalty(hyp_lengths), 1)
 
-            # flatten so we can use top_k
-            scores_flat = tf.reshape(scores, [-1])
+            # reshape to batch_size x (beam_size * vocabulary) for topk
+            input_beam_size = bs_state.input_beam_size
+            scores_flat = tf.reshape(
+                scores, [-1, input_beam_size * len(self.vocabulary)])
 
             # shape(both) = beam
             topk_scores, topk_indices = tf.nn.top_k(
                 scores_flat, self._beam_size)
-
-            topk_scores.set_shape([self._beam_size])
-            topk_indices.set_shape([self._beam_size])
+            topk_indices.set_shape([None, self._beam_size])
+
+            # first, we need offset to flatten topk indices
```

Comment:
tohle chce vysvětlit víc. Co dělá ten range a co s tim potom dělá ten reshape

Reply:
A klidně to vysvětli nějakým ASCII artem.

Reply:
Range dela presne to, co bys cekal (stejne jako pythonovsky range). Ascii art doplnim